File: src/folderIndex.ts
/**
 * src/folderIndex.ts
 *
 * VS Code's producer of the folder edge index (shared/folderIndex.ts): walk a
 * workspace folder's notes, read each one's references (shared/noteLinks.ts),
 * and resolve every reference with the resolver a click on it uses
 * (utils/linkResolver.ts), so a backlink and the file a click opens can never
 * disagree about which note a link means.
 *
 * Pure by the linkResolver precedent: no `vscode` import. The provider lends
 * the IO (the walk, a file's text, the workspace file index) and owns the
 * wiring (who asked, what changed on disk).
 *
 * The work is cached at two grains, because they go stale for different
 * reasons. A note's READING (its references and attributes) changes only when
 * that file does, so a save re-reads one file. A reference's RESOLUTION
 * depends on which files exist and on nothing a save can change, so it is
 * kept until a file is created or deleted; a save then resolves only the
 * references the saved note wrote anew. Resolution is in-memory work over the
 * file list, never a disk read, which is why `isFile` is answered from that
 * list rather than by a stat.
 *
 * The one divergence that choice buys: a target missing from the provider's
 * workspace file index (excluded by `files.exclude`, or past that index's
 * own cap) reads here as unresolved, although a click would stat it and open
 * it. The index errs toward a dangling edge, never toward a wrong one.
 */
import * as path from "path";
import { DOCUMENT_EXTENSIONS } from "../shared/documentExtensions";
import type { FolderEdge, FolderIndex, FolderNode } from "../shared/folderIndex";
import { readNote, type NoteReading } from "../shared/noteLinks";
import { resolveLinkPath, resolveWikiTarget, type ResolverIo } from "./utils/linkResolver";

/**
 * How many notes one walk reads before it stops and says it stopped. Enough
 * for a working vault; a folder past it is indexed partially and the index
 * carries `truncated`, so no view reads the cut as an absence of links.
 */
export const FOLDER_INDEX_CAP = 2000;

/** What the provider lends: everything that touches the disk or the editor. */
export interface FolderIndexIo {
    /** Up to `limit` note files (absolute fsPaths) under `root`, any order. */
    listNotes(root: string, limit: number): Promise<readonly string[]>;
    /** A note's text, or null when it cannot be read. */
    readText(fsPath: string): Promise<string | null>;
    /** Every file in the workspace (absolute fsPaths): what a reference resolves against. */
    fileIndex(): Promise<readonly string[]>;
    /** `birta.smartLinks` for documents under `root`. */
    smartLinks(root: string): boolean;
    /** Let the host's event loop run; called between batches of resolutions. */
    yieldToHost?(): Promise<void>;
}

/** How many fresh resolutions run between two yields to the host. */
const RESOLVE_BATCH = 200;

const NOTE_EXT = new Set(DOCUMENT_EXTENSIONS.map((e) => `.${e}`));

/** The walk's glob: every file this editor opens is a note. */
export const NOTE_GLOB = `**/*.{${DOCUMENT_EXTENSIONS.join(",")}}`;

/** Is this a file the index reads as a note? */
export function isNotePath(fsPath: string): boolean {
    return NOTE_EXT.has(path.extname(fsPath).toLowerCase());
}

function toPosix(p: string): string {
    return p.split(path.sep).join("/");
}

/** `abs` relative to `root` in POSIX form, or null when it is not under `root`. */
function underRoot(root: string, abs: string): string | null {
    const rel = path.relative(root, abs);
    if (rel === "" || rel.startsWith("..") || path.isAbsolute(rel)) { return null; }
    return toPosix(rel);
}

interface RootState {
    /** The walk's note list, sorted; null until walked, dropped when the file set changes. */
    notes: string[] | null;
    truncated: boolean;
    /** Readings per note fsPath; an entry is dropped when its file changes. */
    readings: Map<string, NoteReading | null>;
    /** Resolved targets by (kind, naming directory, target, smartLinks); dropped when the file set changes. */
    resolutions: Map<string, string | null>;
    /** The last assembled index; dropped on any change under the root. */
    index: FolderIndex | null;
    /** In-flight assembly, so two asks at once build once. */
    building: Promise<FolderIndex> | null;
}

export class FolderIndexer {
    private readonly roots = new Map<string, RootState>();

    constructor(private readonly io: FolderIndexIo) {}

    private state(root: string): RootState {
        let s = this.roots.get(root);
        if (!s) {
            s = { notes: null, truncated: false, readings: new Map(), resolutions: new Map(), index: null, building: null };
            this.roots.set(root, s);
        }
        return s;
    }

    /** The roots holding `fsPath`, among those indexed so far. */
    private rootsOf(fsPath: string): RootState[] {
        const out: RootState[] = [];
        for (const [root, s] of this.roots) {
            if (underRoot(root, fsPath) !== null) { out.push(s); }
        }
        return out;
    }

    /**
     * A file under an indexed root was written. Only that note's reading is
     * stale: a write changes which references a note makes, never which files
     * exist, so every resolution already made still holds. Returns whether
     * anything indexed changed, so the caller knows to re-send.
     */
    fileChanged(fsPath: string): boolean {
        if (!isNotePath(fsPath)) { return false; }
        let touched = false;
        for (const s of this.rootsOf(fsPath)) {
            s.readings.delete(fsPath);
            s.index = null;
            touched = true;
        }
        return touched;
    }

    /**
     * A file was created or deleted (what a rename fires too). A root holding
     * it walks again. Its resolutions go too, for any file: a dangling link can
     * have been waiting for exactly that one. A NOTE created or deleted
     * anywhere clears every root's resolutions, because a wikilink resolves
     * against the whole workspace's files and the nearest match can move to
     * another folder.
     */
    fileSetChanged(fsPath: string): boolean {
        let touched = false;
        for (const s of this.rootsOf(fsPath)) {
            s.notes = null;
            s.resolutions.clear();
            s.readings.delete(fsPath);
            s.index = null;
            touched = true;
        }
        if (isNotePath(fsPath)) {
            for (const s of this.roots.values()) {
                if (s.resolutions.size === 0 && s.index === null) { continue; }
                s.resolutions.clear();
                s.index = null;
                touched = true;
            }
        }
        return touched;
    }

    /** The index of `root`, assembled from what is cached and reading only what is not. */
    async indexFor(root: string): Promise<FolderIndex> {
        const s = this.state(root);
        if (s.index) { return s.index; }
        if (s.building) { return s.building; }
        const build = this.assemble(root, s).finally(() => { s.building = null; });
        s.building = build;
        return build;
    }

    private async assemble(root: string, s: RootState): Promise<FolderIndex> {
        if (!s.notes) {
            // One past the cap, so a folder holding exactly the cap is not called truncated.
            const listed = await this.io.listNotes(root, FOLDER_INDEX_CAP + 1);
            const notes = listed.filter(isNotePath).sort();
            s.truncated = notes.length > FOLDER_INDEX_CAP;
            s.notes = notes.slice(0, FOLDER_INDEX_CAP);
        }
        const notes = s.notes;
        await Promise.all(notes.map(async (fsPath) => {
            if (s.readings.has(fsPath)) { return; }
            const text = await this.io.readText(fsPath);
            s.readings.set(fsPath, text === null ? null : readNote(text));
        }));

        // Every note the walk listed exists, whether or not the workspace file
        // index (capped on its own) reached it, so both answer `isFile`.
        const files = [...new Set([...await this.io.fileIndex(), ...notes])];
        const fileSet = new Set(files);
        const resolverIo: ResolverIo = {
            isFile: async (abs) => fileSet.has(abs),
            getFileIndex: async () => files,
        };
        const smartLinks = this.io.smartLinks(root);
        const nodePaths = new Set(notes.map((n) => underRoot(root, n)!));

        // A wikilink resolves by name against the whole folder, so the same
        // name from the same directory always lands on the same note: memoize
        // per directory rather than walking the file list once per link, and
        // keep the answers across assemblies until the file set changes.
        const memo = s.resolutions;
        let resolvedFresh = 0;
        const resolve = async (docFsPath: string, target: string, wiki: boolean): Promise<string | null> => {
            const key = `${wiki ? "w" : "l"}\u0000${smartLinks ? 1 : 0}\u0000${path.dirname(docFsPath)}\u0000${target}`;
            if (memo.has(key)) { return memo.get(key)!; }
            // A cold folder can hold thousands of references; hand the host
            // back its thread between batches rather than holding it for the
            // whole folder.
            if (++resolvedFresh % RESOLVE_BATCH === 0) { await this.io.yieldToHost?.(); }
            const ctx = { docFsPath, workspaceRootFsPath: root, smartLinks };
            const abs = wiki && smartLinks
                ? await resolveWikiTarget(target, ctx, resolverIo)
                : await resolveLinkPath(target, ctx, resolverIo);
            memo.set(key, abs);
            return abs;
        };

        const nodes: FolderNode[] = [];
        const edges: FolderEdge[] = [];
        for (const fsPath of notes) {
            const rel = underRoot(root, fsPath)!;
            const reading = s.readings.get(fsPath) ?? null;
            const stem = path.basename(fsPath, path.extname(fsPath));
            const meta = reading?.meta;
            nodes.push({
                path: rel,
                name: meta?.title ?? stem,
                type: meta?.type ?? null,
                tags: meta?.tags ?? [],
                status: meta?.status ?? null,
                trust: meta?.trust ?? null,
                staleAfter: meta?.staleAfter ?? null,
            });
            for (const link of reading?.links ?? []) {
                if (link.path === "") { continue; }
                const abs = await resolve(fsPath, link.path, link.kind === "wiki");
                const inRoot = abs === null ? null : underRoot(root, abs);
                // A reference that resolves to a file which is not a note (an
                // image, a PDF) is not a relation between notes: it is left out
                // rather than drawn as a dangling note.
                if (inRoot !== null && !isNotePath(abs!)) { continue; }
                edges.push({
                    from: rel,
                    to: inRoot !== null && nodePaths.has(inRoot) ? inRoot : null,
                    target: link.path,
                    kind: link.kind,
                    text: link.text,
                    line: link.line,
                });
            }
        }
        const index: FolderIndex = { rootName: path.basename(root), nodes, edges, truncated: s.truncated };
        s.index = index;
        return index;
    }

    /** `fsPath` as the index names it under `root`, or null when the index does not hold it. */
    selfIn(root: string, fsPath: string, index: FolderIndex): string | null {
        const rel = underRoot(root, fsPath);
        return rel !== null && index.nodes.some((n) => n.path === rel) ? rel : null;
    }
}

File: mac/Sources/BirtaWriterCore/FolderIndex.swift
import Foundation

/// The folder edge index (MAR-467) for a directory window's root: which notes
/// under it name which, as the page's `folderIndex` message carries it
/// (shared/folderIndex.ts). What the Backlinks tab reads.
///
/// The same index the extension's producer builds (`src/folderIndex.ts`),
/// made the same way: walk the root, read each note once (`NoteLinks`), and
/// resolve every reference against the walk's own file list
/// (`NoteLinkResolver`), so a note whose links resolve in VS Code resolves
/// here. Four rules carry over unchanged, and each is a promise the page's
/// views rely on:
///
/// - the walk stops at `cap` notes and says so in `truncated`, so an absent
///   backlink can read as "not reached" rather than "not linked";
/// - an edge is kept only between notes: a reference that resolves to a file
///   which is not a note (an image, a PDF) is dropped, not drawn dangling;
/// - an unresolved reference is kept with `to` nil, because a broken
///   cross-link is not a malformed one;
/// - every path is root-relative and POSIX, like the explorer's.
///
/// The walk is `FileIndex`'s: hidden entries and everything under them are
/// skipped, and package contents are not entered. That is the one place the
/// two producers see different trees, since VS Code's search walks dotted
/// folders its `files.exclude` does not name.
///
/// Built off the main thread and swapped in, like `FileIndex`, by a
/// `FolderIndexer` per root, which the app's `WindowSet` asks again when that
/// root's `DirectoryWatcher` reports a change.
public struct FolderIndex: Equatable, Sendable {
    public struct Node: Equatable, Sendable {
        public var path: String
        public var name: String
        public var type: String?
        public var tags: [String]
        public var status: String?
        public var trust: String?
        public var staleAfter: String?
    }

    public struct Edge: Equatable, Sendable {
        public var from: String
        public var to: String?
        public var target: String
        public var kind: NoteLinks.Kind
        public var text: String
        public var line: Int
    }

    public let rootName: String
    public let nodes: [Node]
    public let edges: [Edge]
    public let truncated: Bool

    /// How many notes one walk reads before it stops and says it stopped:
    /// the extension's `FOLDER_INDEX_CAP`, so the two hosts cut a large
    /// folder at the same size.
    public static let cap = 2000
    /// How many files of any kind the walk keeps to resolve against, the Go
    /// to File index's cap.
    public static let fileCap = 20_000

    /// Is this a file the index reads as a note? Every spelling the editor opens.
    public static func isNotePath(_ path: String) -> Bool {
        let ext = NoteLinkResolver.extname(NoteLinkResolver.basename(path)).lowercased()
        return DocumentTypes.opened.contains { ".\($0)" == ext }
    }

    /// `abs` relative to `root`, or nil when it is not under it (`underRoot`).
    static func relative(_ abs: String, to root: String) -> String? {
        guard abs.hasPrefix(root + "/") else { return nil }
        let rel = String(abs.dropFirst(root.count + 1))
        return rel.isEmpty || rel.hasPrefix("..") ? nil : rel
    }

    /// Resolutions already made, by kind, naming directory and target. Valid
    /// for as long as the file list they were made against: a resolution
    /// depends on which files exist and on nothing a note's text can change.
    public typealias ResolutionMemo = [String: String?]

    /// The index of `root` from what a walk found: `notes` are the note files
    /// it read (absolute), `readings` what each said (absent for one that
    /// could not be read), `files` every file a reference may resolve to.
    ///
    /// Pure, so a test hands it a tree it made up; the walk and the reads are
    /// `FolderIndexer`'s. `memo` carries resolutions between two assemblies
    /// over the same file list, and the caller empties it when the list moves.
    public static func assemble(root: String, notes: [String], readings: [String: NoteLinks.Reading],
                                files: [String], truncated: Bool, memo: inout ResolutionMemo) -> FolderIndex {
        let sorted = notes.sorted { $0.utf16.lexicographicallyPrecedes($1.utf16) }
        // Every note the walk read exists, whatever the file list reached.
        var allFiles = files
        let fileSet = Set(files)
        allFiles.append(contentsOf: sorted.filter { !fileSet.contains($0) })
        // Built on the first miss only: an assembly whose every resolution is
        // memoized never pays for the resolver's per-file keys.
        var resolver: NoteLinkResolver?
        let nodePaths = Set(sorted.compactMap { relative($0, to: root) })

        // A wikilink resolves by name against the whole folder, so the same
        // name from the same directory lands on the same note every time:
        // memoized per naming directory rather than resolved once per link.
        func resolve(_ doc: String, _ target: String, wiki: Bool) -> String? {
            let key = "\(wiki ? "w" : "l")\u{0}\(NoteLinkResolver.dirname(doc))\u{0}\(target)"
            if let known = memo[key] { return known }
            if resolver == nil { resolver = NoteLinkResolver(root: root, files: allFiles) }
            let found = wiki ? resolver!.resolveWiki(target, from: doc) : resolver!.resolveLink(target, from: doc)
            memo[key] = .some(found)
            return found
        }

        var nodes: [Node] = []
        var edges: [Edge] = []
        for note in sorted {
            guard let rel = relative(note, to: root) else { continue }
            let reading = readings[note]
            let base = NoteLinkResolver.basename(note)
            let ext = NoteLinkResolver.extname(base)
            let stem = JSText.string(Array(base.utf16.dropLast(ext.utf16.count)))
            let meta = reading?.meta
            nodes.append(Node(path: rel, name: meta?.title ?? stem, type: meta?.type, tags: meta?.tags ?? [],
                              status: meta?.status, trust: meta?.trust, staleAfter: meta?.staleAfter))
            for link in reading?.links ?? [] where !link.path.isEmpty {
                let abs = resolve(note, link.path, wiki: link.kind == .wiki)
                let inRoot = abs.flatMap { relative($0, to: root) }
                // A reference to a file that is not a note is not a relation
                // between notes: left out, not drawn as a dangling note.
                if let abs, inRoot != nil, !isNotePath(abs) { continue }
                edges.append(Edge(from: rel, to: inRoot.flatMap { nodePaths.contains($0) ? $0 : nil },
                                  target: link.path, kind: link.kind, text: link.text, line: link.line))
            }
        }
        return FolderIndex(rootName: NoteLinkResolver.basename(root), nodes: nodes, edges: edges, truncated: truncated)
    }

    /// `assemble` with nothing carried over.
    public static func assemble(root: String, notes: [String], readings: [String: NoteLinks.Reading],
                                files: [String], truncated: Bool) -> FolderIndex {
        var memo = ResolutionMemo()
        return assemble(root: root, notes: notes, readings: readings, files: files, truncated: truncated, memo: &memo)
    }

    /// `file` as the index names it, or nil when the index holds no such note.
    public func path(of file: URL, root: URL) -> String? {
        guard let rel = DirectoryListing.relativePath(of: file, in: root) else { return nil }
        return nodes.contains { $0.path == rel } ? rel : nil
    }

    /// The wire form, `FolderIndex` in shared/folderIndex.ts.
    public var jsonObject: [String: Any] {
        func orNull(_ s: String?) -> Any { s ?? NSNull() }
        return [
            "rootName": rootName,
            "truncated": truncated,
            "nodes": nodes.map { n -> [String: Any] in
                ["path": n.path, "name": n.name, "type": orNull(n.type), "tags": n.tags,
                 "status": orNull(n.status), "trust": orNull(n.trust), "staleAfter": orNull(n.staleAfter)]
            },
            "edges": edges.map { e -> [String: Any] in
                ["from": e.from, "to": orNull(e.to), "target": e.target, "kind": e.kind.rawValue,
                 "text": e.text, "line": e.line]
            },
        ]
    }
}

/// Builds one root's `FolderIndex`, again and again, reusing what did not change.
///
/// The app writes a note on every pause in the typing, and every write is a
/// change the root's watcher reports, so the index is rebuilt far more often
/// than the folder really moves. The two costs of a build go stale for
/// different reasons, as in the extension's producer: a note's READING only
/// when that file's bytes change, which the walk sees as a new modification
/// date or size; and a reference's RESOLUTION only when the file list does.
/// So a rebuild after a save walks the tree, reads the one note that changed,
/// and answers every resolution from the memo.
///
/// Not thread-safe, and `@unchecked Sendable` only so it can be handed to the
/// one serial queue its owner calls `build` from.
public final class FolderIndexer: @unchecked Sendable {
    public let root: URL
    private let cap: Int
    private let fileCap: Int
    private let read: (String) -> String?

    private struct Stamp: Equatable {
        var modified: Date?
        var size: Int?
    }

    private var readings: [String: (stamp: Stamp, reading: NoteLinks.Reading?)] = [:]
    private var lastFiles: [String]?
    private var memo = FolderIndex.ResolutionMemo()

    /// How many notes the last `build` read from disk rather than from its
    /// cache; what a test asks to see that a save re-reads one file.
    public private(set) var lastReadCount = 0

    public init(root: URL, cap: Int = FolderIndex.cap, fileCap: Int = FolderIndex.fileCap,
                read: @escaping (String) -> String? = { try? String(contentsOfFile: $0, encoding: .utf8) }) {
        self.root = root
        self.cap = cap
        self.fileCap = fileCap
        self.read = read
    }

    /// Walk the root as `FileIndex` does, read what changed, and assemble.
    /// Blocking: the caller runs it off the main thread.
    public func build(fileManager: FileManager = .default) -> FolderIndex {
        let rootPath = root.standardizedFileURL.path
        let keys: [URLResourceKey] = [.isRegularFileKey, .contentModificationDateKey, .fileSizeKey]
        var notes: [(path: String, stamp: Stamp)] = []
        var files: [String] = []
        if let walk = fileManager.enumerator(at: root, includingPropertiesForKeys: keys,
                                             options: [.skipsHiddenFiles, .skipsPackageDescendants]) {
            for case let url as URL in walk {
                guard let values = try? url.resourceValues(forKeys: Set(keys)), values.isRegularFile == true,
                      let rel = DirectoryListing.relativePath(of: url, in: root) else { continue }
                let abs = rootPath + "/" + rel
                if FolderIndex.isNotePath(abs) {
                    notes.append((abs, Stamp(modified: values.contentModificationDate, size: values.fileSize)))
                    // One past the cap, so a folder holding exactly the cap
                    // is not called truncated.
                    if notes.count > cap { break }
                }
                if files.count < fileCap { files.append(abs) }
            }
        }
        let truncated = notes.count > cap
        let kept = notes.sorted { $0.path.utf16.lexicographicallyPrecedes($1.path.utf16) }.prefix(cap)

        var fresh: [String: (stamp: Stamp, reading: NoteLinks.Reading?)] = [:]
        var current: [String: NoteLinks.Reading] = [:]
        lastReadCount = 0
        for note in kept {
            if let cached = readings[note.path], cached.stamp == note.stamp, cached.stamp.modified != nil {
                fresh[note.path] = cached
            } else {
                lastReadCount += 1
                fresh[note.path] = (note.stamp, read(note.path).map(NoteLinks.readNote))
            }
            if let reading = fresh[note.path]?.reading { current[note.path] = reading }
        }
        readings = fresh

        // What the resolver is built from: the file list and the notes, which
        // the list may not hold once it is past its own cap.
        let resolvable = files + kept.map(\.path)
        if resolvable != lastFiles {
            memo.removeAll()
            lastFiles = resolvable
        }
        return FolderIndex.assemble(root: rootPath, notes: kept.map(\.path), readings: current,
                                    files: files, truncated: truncated, memo: &memo)
    }
}
